src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs

```diff
@@ -145,25 +136,35 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
 
         public void Dispose()
         {
-            var disposables = Interlocked.Exchange(ref _disposables, null);
-
-            if (disposables != null)
+            lock (_disposables)
```

Comment:
It's safe to lock on an arbitrary object when other code can't get at it. What you're suggesting is *safer* - but that's not the point. Your safety is not guaranteed.